[vda-storagenode-registry ] Initial version (V 0.1.0)

ITStar10 opened the pull request:
First implementation of https://github.com/verida/VIPs/blob/develop/VIPs/vip-10.md

Closes #118

ITStar10 committed:
Update VDA-Reward-Contract/scripts

ITStar10 committed:
Contract : Created

ITStar10 committed:
[VDA-StorageNode-Registry] Version 0.1.0

- Contract completed
- Test code completed and passed
- `ReadMe.md` added
- `ChangeLog.md` added
- Add multi-sign support

ITStar10 committed:
[VDA-SBT] minor update in Contract file

ITStar10 committed:
Update commit for function parameters

ITStar10 committed:
Update : Gas Optimization

Commit:
Cleanup readme

tahpot commented on VDA-StorageNode-Registry/test/index.test.ts:
```diff
@@ -0,0 +1,939 @@
+
+                let tx = await contract.removeDatacenter(datacenterIds[0])
+                await expect(tx).to.emit(contract, "RemoveDataCenter").withArgs(datacenterIds[0]);
+
```
Add a checkt oconfirm the removed data center can't be fetched with `getDatacenters`

tahpot commented on VDA-StorageNode-Registry/test/index.test.ts:
```diff
@@ -0,0 +1,939 @@
+
+                let tx = await contract.removeDatacenter(datacenterIds[0])
+                await expect(tx).to.emit(contract, "RemoveDataCenter").withArgs(datacenterIds[0]);
+
```
Add a checkt oconfirm the removed data center can't be fetched with `getDatacenters`

ITStar10 replied:
Added.

tahpot commented on VDA-StorageNode-Registry/test/index.test.ts:
```diff
@@ -0,0 +1,939 @@
+            snapShotWithDatacenters = await takeSnapshot();
+        })
+
+        describe("Add storage node", () => {
```
Add check for invalid didAddress.

Add check that `establishmentDate` is set correctly.

Add check that `getNodeByAddress` and `getNodeByEndpoint` return an extra `status` attribute (VIP-10 was updated).

ITStar10 replied:
- Added check for invalid didAddresses. 
- Not able to check the correctness of `establishmentDate`. In the solidity, it set the `block.timestamp` as the `establishmentDate`. This `block.timestamp` is not representing the current time.
- `status` value check added in `getNodeByAddress` & `getNodeByEndpoint()` test

tahpot commented on VDA-StorageNode-Registry/test/index.test.ts:
```diff
@@ -0,0 +1,939 @@
+                );
+            })
+
+            it("Failed: Registered didAddress & endpointURI", async () => {
```
Please rename this test or add more detail, I'm unclear what is eing tested.

ITStar10 replied:
Updated to `Duplicated didAddress and enpointURI`.
Registered `didAddress` or `endpointURI` can't be duplicated.

tahpot commented on VDA-StorageNode-Registry/test/index.test.ts:
```diff
@@ -0,0 +1,939 @@
+
+            })
+
+            it("Failed: Removing didAddress & endpointURI", async () => {
```
Please rename this test or add more detail, I'm unclear what is eing tested.

ITStar10 replied:
DidAddress & endpoints that are in pending removal states can't be used in `addNode()` function.
Renamed to "Failed: didAddress & endpointURI in pending `removal` status".

tahpot commented on VDA-StorageNode-Registry/test/index.test.ts:
```diff
@@ -0,0 +1,939 @@
+                })
+            })
+
+            describe("Get by Country", () => {
```
Is there a solidity limit to how many results this will return?

tahpot commented on VDA-StorageNode-Registry/test/index.test.ts:
```diff
@@ -0,0 +1,939 @@
+                })
+            })
+
+            describe("Get by Region", () => {
```
Is there a solidity limit to how many results this will return?

ITStar10 replied:
Please check following:
- https://betterprogramming.pub/issues-of-returning-arrays-of-dynamic-size-in-solidity-smart-contracts-dd1e54424235
- https://github.com/openethereum/parity-ethereum/issues/6293

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,215 @@
+     * @param nodeInfo Node information to be added
+     * @param requestSignature The request parameters signed by the `didAddress` private key
+     * @param requestProof Used to verify request
+     * @param authSignature Signature signed by a trusted signer
```
Please add this parameter and a description of how it's used to the VIP-10

ITStar10 replied:
Updated : https://github.com/verida/VIPs/pull/5

tahpot commented on VDA-StorageNode-Registry/test/helpers.ts:
```diff
@@ -0,0 +1,126 @@
+    const proofString = `${user.address}${user.address}`.toLowerCase();
+    const requestProof = EncryptionUtils.signData(proofString, privateKeyBuffer);
+
+    const authMsg = ethers.utils.solidityPack(
```
Do we need to consider a `nonce` here to prevent replay attacks of re-adding a node that is removed?

ITStar10 replied:
`nonce` for `authMsg`?

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,545 @@
+     * @param regionCode Unique region string code
+     */
+    function validateRegionCode(string calldata regionCode) internal pure {
+        if (bytes(regionCode).length == 0) {
```
Also check max-length based on values in https://github.com/verida/verida-js/blob/main/packages/account-node/src/countries.ts

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,545 @@
+    function addDatacenter(Datacenter calldata data) external payable onlyOwner override
+         returns(uint) {
+        {
+            if (bytes(data.name).length == 0) {
```
Should we set a max length? 256 characters?

Do we force alpha-numeric with `-` for all `name` and `id` fields?

ITStar10 replied:
No need to set max length.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,545 @@
+            revert InvalidSignature();
+        }
+
+        bytes memory rawMsg = abi.encodePacked(didAddress);
```
Do we need a nonce to prevent re-entry attacks?

ITStar10 replied:
No, need in this AuthSignature verification.
`nonce` check is done in the `verifyRequest()` function.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,545 @@
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function addNode(
```
Needs to support `establishmentDate` (see spec)

ITStar10 replied:
Updated

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,545 @@
+        return _nodeMap[nodeId];
+    }
+
+    function filtrActiveStorageNodes(EnumerableSetUpgradeable.UintSet storage ids) internal view returns(StorageNode[] memory) {
```
Typo, should be `filterActiveStorageNodes`

ITStar10 replied:
Updated.

Commit:
Fix typo

tahpot reviewed (changes requested):
This is a great start!

The quality of the smart contract and tests is really good.

Can you please add the state variables to VIP-10.

ITStar10 committed:
Test done with 3 structs

ITStar10 committed:
Updated : Use 2 structs for StorageNode

ITStar10 committed:
Merge branch 'feature/118-vda-storagenode-registry-create' of github.com:verida/blockchain-contracts into feature/118-vda-storagenode-registry-create

ITStar10 committed:
Minor Update: Uncommented test code

ITStar10 committed:
[0.2.0] Staking feature added

- Add token staking to add/remove nodes
- Test code updated
- Deploy script updated

ITStar10 committed:
[0.2.1] Logging issues and slashing feature added

- `withdraw()` function updated so that any DID can withdraw tokens rewarded by `logNodeIssue()` function
- Added functions for logging and slashing
- Add functions for owner to withdraw tokens staked by `logNodeIssue()` function

ITStar10 committed:
[0.2.1] Update function names

ITStar10 committed:
[0.2.1] Update typos

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @dev Only the contract owner is allowed to call this function
+     * @param isRequired The new value to be updated
+     */
+    function setStakingRequired(bool isRequired) external payable;
```
Why is this payable?

ITStar10 replied:
This is onwer specific function. Declare as payable for gas optimization.
Replied on Discord channel.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @param requestProof Used to verify request
+     * @param authSignature Signature signed by a trusted signer
+     */
+    function addNode(
```
Should this be payable?

ITStar10 replied:
This is not payable function. The users should pay with Verida token, not with native token of the chain.
Payable functions are used to receive the native token of the chain. For example, MATIC in the Polygon, ETH in the Ethereum, and etc.

ITStar10 replied:
Users should approve the Verida token to this `StorageNodeRegistry` contract before they call `addNode()` function.
Of course, the required amount of Verida token  should be approved.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @dev Only the contract owner is allowed to call this function
+     * @param newVal The new value to be updated
+     */
+    function updateStakePerSlot(uint newVal) external payable;
```
Why is this payable?

ITStar10 replied:
This is onwer specific function. Declare as payable for gas optimization.
Replied on Discord channel.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @dev Only the contract owner is allowed to call this function
+     * @param minSlots The new value to be updated
+     */
+    function updateMinSlotCount(uint minSlots) external payable;
```
Why is this payable?

ITStar10 replied:
This is onwer specific function. Declare as payable for gas optimization.
Replied on Discord channel.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @dev Only the contract owner is allowed to call this function
+     * @param maxSlots The new value to be updated
+     */
+    function updateMaxSlotCount(uint maxSlots) external payable;
```
Why is this payable?

ITStar10 replied:
This is onwer specific function. Declare as payable for gas optimization.
Replied on Discord channel.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @dev Only the contract owner is allowed to call this function
+     * @param newTokenAddress New token contract address
+     */
+    function updateTokenAddress(address newTokenAddress) external payable;
```
Why is this payable?

ITStar10 replied:
This is onwer specific function. Declare as payable for gas optimization.
Replied on Discord channel.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @param didAddress DID address
+     * @param tokenAmount Depositing amount of Verida token
+     */
+    function depositToken(address didAddress, uint tokenAmount) external;
```
Does this need payable?

ITStar10 replied:
This is the same as `addNode()` function. Users should approve the Verida token before they call this function.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @notice Update the `NODE_ISSUE_FEE` in the _slotInfo.
+     * @param value New fee value to be set.
+     */
+    function updateNodeIssueFee(uint value) external payable;
```
Why is this payable?

ITStar10 replied:
This is onwer specific function. Declare as payable for gas optimization.
Replied on Discord channel.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+     * @dev Only the contract owner call call this function
+     * @param value Time in seconds unit
+     */
+    function updateSameNodeLogDuration(uint value) external payable;
```
Why is this payable?

ITStar10 replied:
This is onwer specific function. Declare as payable for gas optimization.
Replied on Discord channel.

tahpot commented on VDA-StorageNode-Registry/contracts/IStorageNodeRegistry.sol:
```diff
@@ -0,0 +1,489 @@
+        uint reasonCode,
+        bytes calldata requestSignature,
+        bytes calldata requestProof
+    ) external;
```
Does this need to be payable?

ITStar10 replied:
Same as `addNode()` function. They should approve the Verida token before they call this function.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+        }
+
+        // Release staked token
+        releaseToken(didAddress, tx.origin);
```
Just a thought. Perhaps we don't automatically unstake the tokens when `removeNodeComplete()` and instead they have to be removed via `withdraw()`? In that way it's possible to remove a node and then add another node, without unstaking / restaking.

What do you think?

ITStar10 replied:
Good idea! Here is my thought.
By the way, we can't be sure that the staked amount is enough for the user when they add a node again.
If it's not enough, they should deposit token. This can be by various reasons. For example, the `slotCount` in the user's input might be different from the previous one. Or the contract owner might change the StakePerSlot.
In case, if a user has some staked tokens in the contract, he should calculate how many tokens he should deposit more to add a new node. If there is no staked tokens, the user can calculate the required tokens so simply by `STAKE_PER_SLOT * SLOT_COUNT`. But if he has some staked tokens in the contract, he should calculate required tokens a bit complicated. He should ask the contract how many tokens are remaining and then substract this remaining tokens from the total required tokens. This might be unconveninent for users.

Also, some users don't add node and want to release the staked tokens when they comple remove operation.

And this is another reason. If users should deposit token in `addNode`, it'd be ought to release the token in the `removeNodeComplete`.

tahpot replied:
Thanks for your thoughts.

So you suggest leaving the functionality as it current is?

ITStar10 replied:
Yes, let's leave it as now.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+     * @return StorageNode StoargeNode struct
+     * @return string Status string
+     */
+    function getNodeWithStatus(uint nodeId) private view returns(StorageNode memory, string memory) {
```
Please rename to `getNodeById()`

ITStar10 replied:
This is private function and not used externally.
And this functions returns node and it's status. So named it so. If we rename this function by `getNodebyId()`, it seems like it returs a node that is matched with the inputed nodeId.
Let me know your thougts.

tahpot replied:
Ok, thats makes senses.

Leave as is.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateTokenAddress(address newTokenAddress) external payable onlyOwner {
```
We shouldn't every cahnge the token address.. so is this a security risk?

ITStar10 replied:
This might be a security risk.
Users don't like to change the token address.
Also, there are many users and staked tokens, it might be a bit complicated for us to change the token. Of course we can change the token address simply. After that, we should convert current staked tokens to the new token. This'd be so complicated.

tahpot replied:
Ok. Let's remove the ability to change the token.

ITStar10 replied:
Ok

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+        emit UpdateTokenAddress(oldAddress, newTokenAddress);
+    }
+
+    /**
```
Add comment

ITStar10 replied:
Comment where?

tahpot replied:
Sorry, ignore this.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+            totalAmount = requiredTokenAmount(_nodeMap[nodeId].slotCount);    
+        }
+
+        if (_stakedTokenAmount[didAddress] <= totalAmount) {
```
I think we should return a negative value of there aren't enough tokens staked. That way node operators can know exactly the amount they need to stake to meet their requirements.

ITStar10 replied:
That make sense.
Let me check on my side.

ITStar10 replied:
Updated

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+
+        uint excessAmount = getExcessTokenAmount(didAddress);
+
+        if (excessAmount == 0) {
```
<= =

ITStar10 replied:
Will update after `getExcessTokenAmount()` updated.

ITStar10 replied:
Updated

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+        return _slotInfo.totalIssueFee;
+    }
+
+    /**
```
Add comment

ITStar10 replied:
The comments are in the `IStorageNodeRegistry.sol` file.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+        _slotInfo.MAX_SLOTS = 20000;
+
+        _slotInfo.NODE_ISSUE_FEE = 5 * (10 ** ERC20Upgradeable(tokenAddress).decimals());
+        _slotInfo.SAME_NODE_LOG_DURATION = 1 hours;
```
`Accounts can not log more than 4 issues against any node for a given 24 hour period.`

This requirement seems to be missing?

ITStar10 replied:
Pls check the line#1017 in the `StorageNodeRegistry.sol` file.
I didn't define a constant for this `4` in the contract.

ITStar10 replied:
Updated.
Added `LOG_LIMIT_PER_DAY` variable in the `SlotInfo` struct.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+
+        
+        DIDLogInformation storage logs = _didLogs[didAddress];
+        // Check whether there are 4 logs in 24 horus
```
This should be a variable

ITStar10 replied:
You mean, the number of lgos in 24 hours can be variable?

ITStar10 replied:
Added `LOG_LIMIT_PER_DAY` Variable in `SlotInfo` struct.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+        }
+
+        // Add or update
+        if (logs._issueList.length < 4) {
```
Can you explain this code? Why 4?

ITStar10 replied:
Here the `4` is the available number of logs in the 24 hours.
In that code `logs._issueList` keeps the logs in 24 hours. At first it is an empty array.
So if there are logs less than 4, it pushs the logs in the array.
After the length of array reached 4, it replaced the oldest log with the new one.

tahpot replied:
Ok, can you please add a comment that explains that. Make `4` a variable that owner can change as well.

ITStar10 replied:
Already updated as a variable.

tahpot commented on VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol:
```diff
@@ -0,0 +1,1098 @@
+            unchecked { ++i; }
+        }
+
+        emit Slash(nodeDID, reasonCode, distributeTotalAmount, moreInfoUrl);
```
Also emit how many `didAddress`'s received tokens.

Should we also emit a `loggedIssueDeposit()` event to each `didAddress`? That way a `didAddress` can monitor for events where they've been rewarded for logging an issue?

ITStar10 replied:
We can emit how many DID addresses received tokens.

By the way, it's not conveninen to emit per each address. Emitting an event also requires gas fee.
I think, the users can monitor the logged DID & Reason code to know that they're rewarded. After that, they can know their excess token amounts.

tahpot replied:
Ok, sounds good.

tahpot reviewed (commented):
This is looking great @itstar10.

See my comments inline.

ITStar10 committed:
[0.2.1] Updates on review

- Define `LOG_LIMIT_PER_DAY` state variable instead of constant
- Minor updates on `slash()` function
- Update test code to test log limit per day updates

tahpot reviewed (approved):
PR approved once comment added as per above.

ITStar10 committed:
[0.2.1] Removed the token update feature
